### PR TITLE
deinit solution 

### PR DESCRIPTION
**Pull request by mattmassicotte**
An example of how to deal with synchronous deinit errors. Addresses https://github.com/apple/swift-migration-guide/issues/30.

Depends on #20, but can be seperated out pretty easily.

### CODE REVIEW

File: Guide.docc/CommonProblems.md [hunk 1]
```diff
@@ -719,3 +719,65 @@ class WindowStyler {
...
+class WindowStyler {
+    // A MainActor-isolated type
+    private let store = StyleStore()
+
+    init {
+        // infers MainActor
+    }
+
+    deinit {
+        store.stopNotifications()
```
Comment (ktoso) [inline Guide.docc/CommonProblems.md:740]: note an `// error` here in the code snippet so that it is clear from the snippet itself that it is "bad"

File: Guide.docc/CommonProblems.md [hunk 2]
```diff
@@ -719,3 +719,65 @@ class WindowStyler {
...
+ 6 |     deinit {
+ 7 |         store.stopNotifications()
+   |               `- error: call to main actor-isolated instance method 'stopNotifications()' in a synchronous nonisolated context
+ 8 |     }
+ 9 | }
+```
+
+While this might feel surprising, given that this type is `MainActor`-isolated,
+this is not a new constraint.
+The thread that executes a deinitializer has never been guaranteed and
```
Comment (ktoso) [inline Guide.docc/CommonProblems.md:759]: I'd reword this a bit, we should use this to educate a bit about how deinits work.

```suggestion
Since a deinitializer is synchronous, it is allowed to run inline wherever the last reference to this actor
was released. This is the same as any other reference counted object in Swift works, and deinitializing
the object as soon as possible this way you to have predictable deinitializaiton semantics which are one
of the benefits of automatic reference counting.
```

and let's leave a TODO for us: `deinit isolated` is coming maybe even in 6 this year... so we should then add it here.

File: Guide.docc/CommonProblems.md [hunk 3]
```diff
@@ -719,3 +719,65 @@ class WindowStyler {
...
+it is _critical_ to ensure you do not capture `self`, even implicitly.
+
+```swift
+@MainActor
+class WindowStyler {
+    // A MainActor-isolated type
+    private let store = StyleStore()
+
+    deinit {
+        Task { [store] in
```
Comment (ktoso) [inline Guide.docc/CommonProblems.md:775]: Maybe worth pointing out that Task is NOT main actor isolated
```suggestion
        Task { [store] in // not actor isolated, because since definit is not either
```

Comment (ktoso) [reply Guide.docc/CommonProblems.md:775]: or maybe make a Task.detached here even tbh?

File: Guide.docc/CommonProblems.md [hunk 4]
```diff
@@ -719,3 +719,65 @@ class WindowStyler {
...
+            await store.stopNotifications()
+        }
+    }
+}
+```
+
+> Important: **Never** extend the life-time of `self` from within
+`deinit`. Doing so will crash at runtime.
```
Comment (ktoso) [inline Guide.docc/CommonProblems.md:783]: Good!

**Review by ktoso — APPROVED** [review]

File: Guide.docc/CommonProblems.md (hunk 2 above)
Comment (mattmassicotte) [reply Guide.docc/CommonProblems.md:759]: https://github.com/apple/swift-migration-guide/issues/55

File: Guide.docc/CommonProblems.md (hunk 3 above)
Comment (mattmassicotte) [reply Guide.docc/CommonProblems.md:775]: I bet making this clearer with a comment will be easier for most readers to digest than using `Task.detached`, even if that is more explicit semantically.

File: Guide.docc/CommonProblems.md [hunk 5]
```diff
@@ -719,3 +719,65 @@ class WindowStyler {
 All `Sendable` properties can still be safely accessed in this `init` method.
 And while any non-`Sendable` properties cannot,
 they can still be initialized by using default expressions.
+
+### Non-Isolated Deinitialization
+
+By default, any custom initializers you write will have the same isolation as
+their enclosing scope.
```
[the remark this answers is not in the document]
Comment (mattmassicotte) [reply Guide.docc/CommonProblems.md]: I'm going to mark this down for discussion with @hborla. I really appreciate you bringing it up.
